### ISSUE TITLE
Possible race condition on multiprocessing.Manager().dict() on macOS

### ISSUE THREAD

**Issue by 7e477b84-0ada-4db8-9e26-70bff58e8287** [python/cpython#87934]
BPO | [43768](https://bugs.python.org/issue43768)
--- | :---
Nosy | @ronaldoussoren, @pitrou, @ned-deily, @applio

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-04-07.19:12:33.180>
labels = ['OS-mac', 'type-bug', '3.9']
title = 'Possible race condition on multiprocessing.Manager().dict() on macOS'
updated_at = <Date 2021-04-17.00:45:40.905>
user = 'https://bugs.python.org/jerryc05'
```

bugs.python.org fields:
```python
activity = <Date 2021-04-17.00:45:40.905>
actor = 'ned.deily'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['macOS']
creation = <Date 2021-04-07.19:12:33.180>
creator = 'jerryc05'
dependencies = []
files = []
hgrepos = []
issue_num = 43768
keywords = []
message_count = 1.0
messages = ['390468']
nosy_count = 5.0
nosy_names = ['ronaldoussoren', 'pitrou', 'ned.deily', 'davin', 'jerryc05']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue43768'
versions = ['Python 3.9']
```

</p></details>

**Comment by ronaldoussoren** [comment]
I tested the script on my machine (macOS 13.0.1, python 3.9, 3.10 and 3.11 all installed using the python.org installer), and the error occurs intermittently, likewise with a fresh build of 3.12.   Disabling the local firewall does not avoid this problem.

This appears to be a timing problem, the main process is not yet listening to the socket when the child tries to connect.

Below is a crude hack that implements a retry loop and appears to fix the issue for me. Added as an inline patch instead of a PR because I'm far from convinced that this is a correct fix. I've barely used multiprocessing myself and know to little about its design to know what the correct place would be to implement a retry loop.

```patch
diff --git a/Lib/multiprocessing/connection.py b/Lib/multiprocessing/connection.py
index b08144f7a1..7954fefd62 100644
--- a/Lib/multiprocessing/connection.py
+++ b/Lib/multiprocessing/connection.py
@@ -625,9 +625,13 @@ def SocketClient(address):
     '''
     family = address_type(address)
     with socket.socket( getattr(socket, family) ) as s:
-        s.setblocking(True)
-        s.connect(address)
-        return Connection(s.detach())
+        for _ in range(3):
+            try:
+                s.setblocking(True)
+                s.connect(address)
+                return Connection(s.detach())
+            except socket.error:
+                time.sleep(0.5)
```

**Comment by ronaldoussoren** [comment]
@applio and/or @gpshead, what would be the correct place to implement a retry loop as sketched in my previous message? Or is retrying not the right solution here?

**Comment by ronaldoussoren** [comment]
This is the same problem as #101225, but with a different limit in the backlog.

**Comment by ronaldoussoren** [comment]
The race condition doesn't happen for me with the fix for #101225. That's technically just reducing the size of window where the race condition can happen, but should be fine given that I've increased the backlog far beyond what's needed to avoid hitting the race (famous last words...)
